### DISCUSSION TITLE
book1 chapter3, vec3, what is the difference of "double operator[](int i) const {return e[i];}" and "double& operator[](int i) {return e[i];}" ?

### DISCUSSION THREAD

**Discussion by jia-heng** [RayTracing/raytracing.github.io#1477] [Q&A] [answered]
double operator[](int i) const { return e[i]; }
double& operator[](int i) { return e[i]; }

Do these two sentences have the same effect?

**Comment by hollasch** [comment]
> `double operator[](int i) const { return e[i]; }`

The `const` following the parameter list (before the function body) indicates that this method is available for constant `vec3` objects. It will return a double-precision floating-point _value_.

> `double& operator[](int i) { return e[i]; }`

This one returns a _reference_ (pointer) to the actual double-precision floating-point value inside the `vec3`. This reference can be used to mutate (change) the `vec3`, so it is **not** available to use on constant `vec3` objects.

**Reply by hollasch** [reply, marked as answer]
I'm afraid this is some C++ low-level detail that's unfortunately necessary here. We really try to keep C++ specifics to a minimum. This could look quite different in your own code, depending on your language of choice.

To be a bit more concrete, the first form lets you grab elements out of any `vec3` object, like

```
vec3 v = foo();
double x = v[0];  // Deconstruct the vector v
double y = v[1];
double z = v[2];
```

The second form lets you change/set vector values, like so:

```
vec3 v;
v[0] = 0.0;
v[1] = 10.0;
v[2] = 100.0;
```

**Reply by jia-heng** [reply]
I got it, thx so much
